### PR TITLE
benchmark with yandex/clickhouse-jdbc

### PR DESCRIPTION
**Pull request by sundy-li**

### CODE REVIEW

File: src/test/java/com/github/housepower/jdbc/benchmark/ProfileBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    public void profileSumDoubleNative() throws Exception {
+        withConnection(benchSelectSumDouble, ConnectionType.NATIVE);
+    }
+
+    @Test
+    public void profileSumDoubleHttp() throws Exception {
+        withConnection(benchSelectSumDouble, ConnectionType.HTTP);
+    }
+}
+
```
Comment (sundy-li) [inline src/test/java/com/github/housepower/jdbc/benchmark/ProfileBenchmark.java]: #70  I got the following benchmark result by using Native-JDBC v1.6 on my macbook
```
Native-sumInt

 0% Total memory (MB):   123, used memory (MB):    18, time elapsed:      0(+     0 ms)
 10% Total memory (MB):   228, used memory (MB):     9, time elapsed:    945(+   945 ms)
 20% Total memory (MB):   358, used memory (MB):     9, time elapsed:   1397(+   452 ms)
 30% Total memory (MB):   514, used memory (MB):     9, time elapsed:   2508(+  1111 ms)
 40% Total memory (MB):   515, used memory (MB):     9, time elapsed:   3180(+   672 ms)
 50% Total memory (MB):   515, used memory (MB):     9, time elapsed:   3684(+   504 ms)
 60% Total memory (MB):   515, used memory (MB):     9, time elapsed:   3977(+   293 ms)
 70% Total memory (MB):   514, used memory (MB):     9, time elapsed:   4360(+   383 ms)
 80% Total memory (MB):   515, used memory (MB):     9, time elapsed:   4688(+   328 ms)
 90% Total memory (MB):   515, used memory (MB):     9, time elapsed:   5122(+   434 ms)
100% Total memory (MB):   515, used memory (MB):     7, time elapsed:   5414(+   292 ms)
49999995000000

HTTP-sumInt

 0% Total memory (MB):   123, used memory (MB):    31, time elapsed:      0(+     0 ms)
 10% Total memory (MB):   221, used memory (MB):     3, time elapsed:    676(+   676 ms)
 20% Total memory (MB):   351, used memory (MB):     3, time elapsed:   1322(+   646 ms)
 30% Total memory (MB):   351, used memory (MB):     3, time elapsed:   1812(+   490 ms)
 40% Total memory (MB):   351, used memory (MB):     3, time elapsed:   2385(+   573 ms)
 50% Total memory (MB):   351, used memory (MB):     3, time elapsed:   2853(+   468 ms)
 60% Total memory (MB):   351, used memory (MB):     3, time elapsed:   3299(+   446 ms)
 70% Total memory (MB):   351, used memory (MB):     3, time elapsed:   3756(+   457 ms)
 80% Total memory (MB):   350, used memory (MB):     3, time elapsed:   4157(+   401 ms)
 90% Total memory (MB):   351, used memory (MB):     3, time elapsed:   4618(+   461 ms)
100% Total memory (MB):   351, used memory (MB):     2, time elapsed:   5012(+   394 ms)
49999995000000


Native-sumDouble

0% Total memory (MB):   123, used memory (MB):    18, time elapsed:      0(+     0 ms)
 10% Total memory (MB):   223, used memory (MB):     6, time elapsed:    783(+   783 ms)
 20% Total memory (MB):   354, used memory (MB):     6, time elapsed:   1499(+   716 ms)
 30% Total memory (MB):   354, used memory (MB):     6, time elapsed:   2267(+   768 ms)
 40% Total memory (MB):   354, used memory (MB):     6, time elapsed:   2932(+   665 ms)
 50% Total memory (MB):   354, used memory (MB):     6, time elapsed:   3602(+   670 ms)
 60% Total memory (MB):   354, used memory (MB):     6, time elapsed:   4276(+   674 ms)
 70% Total memory (MB):   348, used memory (MB):     6, time elapsed:   4962(+   686 ms)
 80% Total memory (MB):   354, used memory (MB):     6, time elapsed:   5660(+   698 ms)
 90% Total memory (MB):   348, used memory (MB):     6, time elapsed:   6329(+   669 ms)
100% Total memory (MB):   353, used memory (MB):     5, time elapsed:   6977(+   648 ms)
6.101578795396764E-5

HTTP-sumDouble

  0% Total memory (MB):   123, used memory (MB):    31, time elapsed:      0(+     0 ms)
 10% Total memory (MB):   507, used memory (MB):     3, time elapsed:   3000(+  3000 ms)
 20% Total memory (MB):   484, used memory (MB):     3, time elapsed:   5429(+  2429 ms)
 30% Total memory (MB):   466, used memory (MB):     3, time elapsed:   7596(+  2167 ms)
 40% Total memory (MB):   431, used memory (MB):     3, time elapsed:   9632(+  2036 ms)
 50% Total memory (MB):   402, used memory (MB):     3, time elapsed:  11799(+  2167 ms)
 60% Total memory (MB):   374, used memory (MB):     3, time elapsed:  14050(+  2251 ms)
 70% Total memory (MB):   350, used memory (MB):     3, time elapsed:  16155(+  2105 ms)
 80% Total memory (MB):   327, used memory (MB):     3, time elapsed:  18369(+  2214 ms)
 90% Total memory (MB):   297, used memory (MB):     3, time elapsed:  20651(+  2282 ms)
100% Total memory (MB):   271, used memory (MB):     2, time elapsed:  23102(+  2451 ms)
4.371712365763836E-5

```

Comment (sundy-li) [reply src/test/java/com/github/housepower/jdbc/benchmark/ProfileBenchmark.java]: https://github.com/yandex/ClickHouse/issues/4002

**Comment by zhang2014** [comment]
Can you update the latest test result？

**Comment by sundy-li** [comment]
> Can you update the latest test result？
I just update the code, now seems the width column of String results to native-jdbc very slow(GC mostly). On other comparations, native-jdbc is significantly faster than http-jdbc.
here's a screenshot on my local machine.
image

The code is in the [benchmark](https://github.com/housepower/ClickHouse-Native-JDBC/blob/feature%2Fbenchmark/src/test/java/com/github/housepower/jdbc/benchmark/WideColumnInsertIBenchmark.java) dirs.

I'll continue to improve the width-String batch insert performance.

**Comment by zhang2014** [comment]
> now seems the wide column of String results to native-jdbc very slow(GC mostly)

I have an idea that if we parse the data and post the data as an external table(we only need to assemble the IO buffer to specified size for each JDBC set of the user and send it directly to the server), and when the user call the `execute` method we send the last empty block for the query to execute.

For this implementation, JDBC only requires a fixed amount of memory to maintain the IO buffer which can be off heap(to avoid gc, which may not be off heap), but the data will be stored as temporary table on the server.

**Pull request merged**